storage/src/index/mod.rs

```diff
@@ -31,967 +31,1047 @@ mod tests {
 
     #[test_traced]
     fn test_index_basic() {
-        let context = deterministic::Context::default();
-        let mut index = Index::init(context.clone(), TwoCap);
-        assert!(context.encode().contains("keys 0"));
-        assert!(context.encode().contains("items 0"));
-
-        // Generate a collision and check metrics to make sure it's captured
-        let key = b"duplicate".as_slice();
-        index.insert(key, 1);
-        index.insert(key, 2);
-        index.insert(key, 3);
-        assert!(context.encode().contains("keys 1"));
-        assert!(context.encode().contains("items 3"));
-
-        // Check that the values are in the correct order
-        assert_eq!(index.get(key).copied().collect::<Vec<_>>(), vec![1, 3, 2]);
-
-        // Ensure cursor terminates
-        {
-            let mut cursor = index.get_mut(key).unwrap();
-            assert_eq!(*cursor.next().unwrap(), 1);
-            assert_eq!(*cursor.next().unwrap(), 3);
-            assert_eq!(*cursor.next().unwrap(), 2);
-            assert!(cursor.next().is_none());
-        }
-
-        // Make sure we can remove keys with a predicate
-        index.insert(key, 3);
-        index.insert(key, 4);
-        index.prune(key, |i| *i == 3);
-        assert_eq!(index.get(key).copied().collect::<Vec<_>>(), vec![1, 4, 2]);
-        index.prune(key, |_| true);
-        // Try removing all of a keys values.
-        assert_eq!(
-            index.get(key).copied().collect::<Vec<_>>(),
-            Vec::<u64>::new()
-        );
-        assert!(context.encode().contains("keys 0"));
-        assert!(context.encode().contains("items 0"));
+        let runner = deterministic::Runner::default();
+        runner.start(|context| async move {
+            let mut index = Index::init(context.clone(), TwoCap);
```

Comment:
Given this only requires `Metrics`, it may be nice to implement a `dummy` default of `Metrics` that doesn't require the whole runner?

Providing a default `Context`, however, is very error-prone and should not be the path we take.